emcc.py

```diff
@@ -3421,18 +3427,18 @@ def modularize():
       script_url = "typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined"
       if shared.Settings.target_environment_may_be('node'):
         script_url_node = "if (typeof __filename !== 'undefined') _scriptDir = _scriptDir || __filename;"
+    lines = ['var _scriptDir = %s;' % script_url, script_url_node]
+    if shared.Settings.CACHE_MODULARIZE_MODULE:
+      # when we do sync compilation we can cache the compiled module to save
+      # time on later factory invocations
+      lines.append('var cachedModule;')
+    lines.append('return (%s);' % src)
     src = '''
 var %(EXPORT_NAME)s = (function() {
-  var _scriptDir = %(script_url)s;
-  %(script_url_node)s
-  return (%(src)s);
+  %(lines)s
 })();
-''' % {
-      'EXPORT_NAME': shared.Settings.EXPORT_NAME,
-      'script_url': script_url,
-      'script_url_node': script_url_node,
-      'src': src
-    }
+''' % {'EXPORT_NAME': shared.Settings.EXPORT_NAME,
+       'lines': '\n  '.join(lines)}
```

Comment:
I wonder if this function could be moved to a library JS file where the `if CACHE_MODULARIZE_MODULE` pattern might be more readable?

Reply:
+1, this is hard to follow. Or maybe just adding the `var cachedModule;` statement using string substitution instead of introducing the array and having to follow the appending and joining.

Reply:
This is absolutely bikeshedding, but I do still think it's clearer as string substitutions.

Reply:
Fair enough. Maybe the non-bikeshedding part can decide it, then: is there a clean way to avoid extra empty lines with string substitutions?

Reply:
Can't think of anything aside from putting the newline character in the substituted string

```python
  src = '''
var %(EXPORT_NAME)s = (function() {	var %(EXPORT_NAME)s = (function() {
  var _scriptDir = %(script_url)s;	  %(lines)s
  %(script_url_node)s%(cached_module_var)s
  return (%(src)s);	
})();	})();
''' % {
      'EXPORT_NAME': shared.Settings.EXPORT_NAME,
      'script_url': script_url,	
      'script_url_node': script_url_node,	
      'cached_module_var': '\nvar cachedModule;' if shared.Settings.CACHE_MODULARIZE_MODULE else '',
      'src': src	
    }
```